---
title: "White Rose DTP - Analytics 2"
author: "Module Coordinator Mark Dunning"
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
    css: stylesheets/styles.css
editor_options: 
  chunk_output_type: inline
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,fig.width = 12,message=FALSE,warning=FALSE)
library(dplyr)

```



# Learning outcomes

- Using annotation databases to map between gene identifers
- Construction and interpretation of common visualisations for RNA-seq
    + scatter plots
    + volcano plots
    + MA-plots 
    + heatmaps
- Customisation of plots

We can now have a list of genes ordered according to their evidence for being differentially-expressed. You should have saved a `de_cellType.rds` object in the previous session. 

```{r}
library(DESeq2)
de_cellType <- readRDS("Robjects/de_cellType.rds")
results_cellType <- results(de_cellType) %>% 
  as.data.frame %>% 
  tibble::rownames_to_column("ENSEMBL")
```

We would love to share this list with our collaborators, or search for our favourite gene in the results. However, the results are not very useful in there current form as each row is named according to an *Ensembl* identifier. Whilst gene symbols are problematic and can change over time, they are the names that are most recognisable and make the results easier to navigate.

# Adding annotation to the DESeq2 results

There are a number of ways to add annotation, but we will demonstrate how to do this using the *org.Mm.eg.db* package. This package is one of several *organism-level* packages in Bioconductor that are re-built every 6 months. These packages are listed on the [annotation section](http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData) of the Bioconductor, and are installed in the same way as regular Bioconductor packages. An alternative approach is to use `biomaRt`, an interface to the [BioMart](http://www.biomart.org/) resource. BioMart is much more comprehensive, but the organism packages fit better into the Bioconductor workflow.


```{r eval=FALSE}
### Only execute when you need to install the package
install.packages("BiocManager")
BiocManager::install("org.Mm.eg.db")
# For Human
BiocManager::install("org.Hs.eg.db")
```

The packages are larger in size that Bioconductor software pacakges, but essentially they are databases that can be used to make *offline* queries. 

```{r message=FALSE}
library(org.Mm.eg.db)
```


First we need to decide what information we want. In order to see what we can extract we can run the `columns` function on the annotation database.

```{r}
columns(org.Mm.eg.db)
```

We are going to filter the database by a key or set of keys in order to extract the information we want. Valid names for the key can be retrieved with the `keytypes` function.

```{r}
keytypes(org.Mm.eg.db)
```

We should see `ENSEMBL`, which is the type of key we are going to use in this case. If we are unsure what values are acceptable for the key, we can check what keys are valid with `keys`

```{r}
keys(org.Mm.eg.db, keytype="ENSEMBL")[1:10]
```



For the top gene in our analysis the call to the function would be:-

```{r eval=FALSE}
select(org.Mm.eg.db, keys="ENSMUSG00000027750",
       keytype = "ENSEMBL",columns=c("SYMBOL","GENENAME")
)
```

Unfortunately, the authors of `dplyr` and `AnnotationDbi` have both decided to use the name `select` in their packages. To avoid confusion and errors, the following code is sometimes used:-

```{r}
AnnotationDbi::select(org.Mm.eg.db, keys="ENSMUSG00000000381",keytype = "ENSEMBL",columns=c("SYMBOL","GENENAME"))
```


To annotate our results, we definitely want gene symbols and perhaps the full gene name. Let's build up our annotation information into a new data frame using the `select` function.

```{r}
anno <- AnnotationDbi::select(org.Mm.eg.db,keys=results_cellType$ENSEMBL,
              columns=c("SYMBOL","GENENAME"),
              keytype="ENSEMBL")
# Have a look at the annotation
head(anno)
```

However, we have a problem that the resulting data frame has more rows than our results table. This is due to the *one-to-many* relationships that often occur when mapping between various identifiers.

```{r}
dim(anno)
dim(results_cellType)
```

Such duplicated entries can be identified using the `duplicated` function. 

```{r}
dup_ids <- anno$ENSEMBL[duplicated(anno$ENSEMBL)]
filter(anno, ENSEMBL %in% dup_ids) %>% 
  arrange(ENSEMBL) %>% head
```

Fortunately, there are not too many so hopefully we won't lose too much information if we discard the entries that are duplicated. The first occurrence of the duplicated ID will still be included in the table.

```{r}
  anno <- AnnotationDbi::select(org.Mm.eg.db,keys=results_cellType$ENSEMBL,
              columns=c("ENSEMBL","SYMBOL","GENENAME","ENTREZID"),
              keytype="ENSEMBL") %>% 
  filter(!duplicated(ENSEMBL))
dim(anno)
```


We can bind in the annotation information to the `results` data frame. 

```{r}
results_annotated <- left_join(results_cellType, anno,by="ENSEMBL") %>% 
  arrange(padj)
head(results_annotated)
```


We can save the results table using the `write.csv` function, which writes the results out to a csv file that you can open in excel.

```{r}
write.csv(results_annotated,file="cellType_DESeq_annotated.csv",row.names=FALSE)
saveRDS(results_annotated, file="Robjects/results_cellType_annotated.rds")
```


# Visualisation

Now we have annotated our results, we can start to explore some common visualisation techniques.

A useful sanity check is to examine the counts for a single gene across biological groups. A simple function for making this plot is `plotCounts`, which normalizes counts by sequencing depth and adds a "pseudocount" of 1/2 to allow for log scale plotting. The counts are grouped by the variables in  `intgroup`, where more than one variable can be specified. Here we specify the gene which had the smallest p-value from the results table created above. You can select the gene to plot by rowname or by numeric index:-

As the results table no longer contains read counts, we will have to load these from a previous object.

```{r}
library(DESeq2)
## Read the counts that we produced previously
dds <- readRDS("Robjects/dds.rds")

plotCounts(dds, "ENSMUSG00000027750",intgroup = "CellType")
```


## Exercise

<div class="exercise">

- Use `plotCounts` with the option `returnData=TRUE`  to get a data frame containing the counts of `ENSMUSG00000027750` in the different cell types. Visualise these data using `ggplot2` and colour the samples appropriately
    + HINT: To get the counts on the same scale as displayed by the plotCounts function you will need to add adjust the scale on the y-axis 
</div>



## The Volcano Plot

A common plot for displaying the results of a differential expression analysis is a *volcano plot*. It is a *scatter plot* that shows statistical significance and the magnitude of difference between conditions. They are used to identify which genes are the most significant and are also changing by the most amount.

The data we need for the plot is contained in our `results_annotated` data frame. This basic plot displays a point for every gene, but does not take advantage of some of the other columns in the data frame.

```{r}
library(ggplot2)
results_annotated %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) + geom_point()
```


One modification is to colour the points according to whether each gene is significant in the analysis. The indicator of significance can be a new column in the data frame that we create on-the-fly using the pipe operator.

```{r}
results_annotated %>% 
  mutate(Significant = padj < 0.05 & abs(log2FoldChange) > 2) %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj), col=Significant)) + geom_point()
```


We can also add the gene names to the plot. This should be straightforward as ggplot2 has a `label` aesthetic that can be mapped to columns in a data frame. The `geom_text` plot will then display the labels. However, the following plot is a bit crowded.

```{r}
## Not a good idea to run this!!
results_annotated %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj), label=SYMBOL)) + geom_point() + geom_text()
```


The problem here is that ggplot2 is trying to label every point with a name; not quite what we want. The trick is to create a label that is blank for most genes and only labels the points we are interested in. The `ifelse` function in R is a convenient way to set the entries in a vector based on a *logical* expression. In this case, make the values in `Label` the same as the gene symbol if the gene is in our list of "top genes". Otherwise, points get labeled with a blank string `""`.

For clarity, we also make the points slightly transparent and use a different colour for the text.

```{r}
N <- 10
top_genes <- results_annotated$ENSEMBL[1:N]
results_annotated %>% 
  mutate(Label = ifelse(ENSEMBL %in% top_genes, SYMBOL, "")) %>%  
  ggplot(aes(x = log2FoldChange, y = -log10(padj), label=Label)) + geom_point(alpha=0.4) + geom_text(col="blue")
```

Finally, a slightly better positioning of text is given by the `ggrepel` package.

```{r}
if(!require(ggrepel)) install.packages("ggrepel")

results_annotated %>% 
  mutate(Label = ifelse(ENSEMBL %in% top_genes, SYMBOL, "")) %>%  
  ggplot(aes(x = log2FoldChange, y = -log10(padj), label=Label)) + geom_point(alpha=0.4) + geom_text_repel(col="blue")
```


## Heatmaps

You may have already seen the use of a heatmap as a quality assessment tool to visualise the relationship between samples in an experiment. Another common use-case for such a plot is to visualise the results of a differential expression analysis.

Here we will take the top 10 genes from the differential expression analysis and produce a heatmap with the `pheatmap` package. We can take advantage of the fact the our counts table contains Ensembl gene names in the rows. Standard subset operations in R can then be used.

The default colour palette goes from low expression in blue to high expression in red, which is a good alternative to the traditional red/green heatmaps which are not suitable for those with forms of colour-blindness.

The counts we are visualising are the *variance-stablised* counts, which are more appropriate for visualisation.

Although `ggplot2` has a `geom_tile` function to make heatmaps, specialised packages such as `pheatmaps` offer more functionality such as clustering the samples.

```{r}
# pheatmap is a specialised package to make heatmaps
library(pheatmap)
top_genes <- results_annotated$ENSEMBL[1:10]
vsd <- vst(dds)
pheatmap(assay(vsd)[top_genes,])
```

The heatmap is more informative if we add colours underneath the sample dendrogram to indicate which sample group each sample belongs to. This we can do by creating a data frame containing metadata for each of the samples in our dataset. With the `DESeq2` workflow we have already created such a data frame. We have to make sure the the rownames of the data frame are the same as the column names of the counts matrix.

```{r}
sampleInfo <- as.data.frame(colData(dds)[,c("Status","CellType")])

pheatmap(assay(vsd)[top_genes,],
         annotation_col = sampleInfo)
```

Any plot we create in RStudio can be saved as a png or pdf file. We use the `png` or `pdf` function to create a file for the plot to be saved into and run the rest of the code as normal. The plot does not get displayed in RStudio, but printed to the specified file. 

```{r}
png("heatmap_top10_genes.png",width=800,height=800)
pheatmap(assay(vsd)[top_genes,],
         annotation_col = sampleInfo)
# dev.off()
```

# Exercise

<div class="exercise">


- Repeat the same heatmap as above, but for the top 100 most differentially-expressed genes **between lactation and virgin**
    + Load your `de_status.rds` file from last week. 
    + Re-arrange by adjusted p-value and join with the annotation table from `org.Mm.eg.db`
    + identify the top 100 genes by p-value and create the heatmap
- Save the plot to a pdf file

</div>





# Annotation with the biomaRt resource

The Bioconductor package have the convenience of being able to make queries offline. However, they are only available for certain organisms. If your organism does not have an `org.XX.eg.db` package listed on the Bioconductor annotation page (http://bioconductor.org/packages/release/BiocViews.html#___AnnotationData), an alternative is to use biomaRt which provides an interface to the popular biomart annotation resource. 

The first step is to find the name of a database that you want to connect to.

```{r}
library(biomaRt)
listMarts()
ensembl=useMart("ENSEMBL_MART_ENSEMBL")
# list the available datasets (species). Replace mouse with the name of your organism
listDatasets(ensembl) %>% filter(grepl("Mouse",description))
```

```{r}
ensembl = useDataset("mmusculus_gene_ensembl", mart=ensembl)
```

Queries to `biomaRt` are constructed in a similar way to the queries we performed with the `org.Mm.eg.db` package. Instead of `keys` we have `filters`, and instead of `columns` we have attributes. The list of acceptable values is much more comprehensive that for the `org.Mm.eg.db` package.

```{r}
listFilters(ensembl) %>% 
    filter(grepl("ensembl",name))
```


```{r eval=FALSE}
listAttributes(ensembl) %>% 
    filter(grepl("gene",name))
```

An advantage over the `org..` packages is that positional information can be retrieved

```{r}
attributeNames <- c('ensembl_gene_id', 'entrezgene_id', 'external_gene_name', "chromosome_name","start_position","end_position")

getBM(attributes = attributeNames,
      filters = "ensembl_gene_id",
      values=top_genes,
      mart=ensembl)
```

# Exercise for next time

<div class="exercise">
- When visualising the expression of `ENSMUSG00000000381` we can also observe an effect of `CellType` on the expression level. 
```{r}
dds$Status <- as.factor(dds$Status)
plotCounts(dds, "ENSMUSG00000000381",intgroup = "Status")
```
- Modify the code producing the plot so that the points on the plot can be identified by `CellType` and `Status` (e.g using different colours and shapes or a facet). You will need to look at the options for `plotCounts` and how to specify multiple groups in `intgroup`.
- Suppose we are interested in the Gene Ontology term "mitotic cell cycle arrest" - `GO:0071850`. What genes belong to this pathway in Mouse? Find out the `ENSEMBL` IDs for these genes are make a heatmap. Does it look like genes in this pathway are differentially-expressed between cell types?


</div>
