---
title: "White Rose DTP - Analytics 2"
author: "Module Coordinator Mark Dunning"
output: 
  html_notebook: 
    toc: yes
    toc_float: yes
    css: stylesheets/styles.css
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



# Learning Objectives

- Which statistical tests are appropriate for RNA-seq data
- Using the DESeq2 package to detect differential expression
- Using a venn diagram to compare gene lists

# Differential expression with `DESeq2`

Now that we are happy that we have normalised the data and that the quality looks good, we can continue to testing for differentially expressed genes. There are a number of packages to analyse RNA-Seq data. Most people use `DESeq2` or `edgeR`. We will use `DESeq2` for the rest of this practical.

First make sure we have all the objects and libraries loaded.



```{r warning=FALSE,message=FALSE}
library(DESeq2)
library(tximport)
```


### Recap of pre-processing

The previous section walked-through the pre-processing and transformation of the count data. Here, for completeness, we list the minimal steps required to process the data prior to differential expression analysis.

Note that although we spent some time looking at the quality of our data , these steps are not required prior to performing differential expression so are not shown here. Remember, `DESeq2` [requires raw counts](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts) so the `vst` transformation is not shown as part of this basic protocol.

```{r message=FALSE}
library(tximport)
library(DESeq2)
dirs <- list.files(file.path("salmon_quant/"))
quant_files <- paste0("salmon_quant/",dirs,"/quant.sf.gz")
names(quant_files) <- dirs

tx2gene <- read.csv("tx2gene.csv")

txi <- tximport(quant_files,type="salmon",tx2gene = tx2gene,ignoreTxVersion = TRUE)

sampleinfo <- read.delim("meta_data/sampleInfo_corrected.txt")
rownames(sampleinfo) <- sampleinfo$run

dds <- DESeqDataSetFromTximport(txi, 
                                colData = sampleinfo,
                                design <- ~CellType)

```

It would be a good idea to save the results of the pre-processing so we don't have to repeat it every time.

```{r}
dir.create("Robjects/",showWarnings = FALSE)
saveRDS(dds, file="Robjects/dds.rds")
```


We will be using these raw counts throughout the workshop and transforming them using methods in the `DESeq2` package. If you want to know about alternative methods for count normalisation they are covered on [this page](https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html).

# Differential expression with DESeq2

We have previously defined the test condition using the `design` argument when we created the object. This can be checked using the `design` function.

Typically we decide the design for the analysis when we create the DESeq2 objects, but it can be modified prior to the differential expression analysis. The design tells `DESeq2` which sample groups to compare in the differential analysis. The name specified must correspond to a column in the sample information.

```{r}
colData(dds)
design(dds) <- ~CellType
```

As we will see, the `DESeq2` package provides a convenient function for performing the differential expression.

The `DESeq` function runs a couple of processing steps automatically to adjust for different library size and gene-wise variability, which you can read about in the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#the-deseq2-model). 


The counts that we have obtained via sequencing are subject to random sources of variation. The purpose of differential expression is to determine if potential sources of biological variation (e.g. counts observed from different sample groups) are greater than random noise.

The median of ratios normalisation method is employed in DESeq2 to account for *sequencing depth* and *RNA composition*. Let's go through a short worked example (courtesy of [https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html](here)) to explain the process.

```{r}
## create a small example matrix of "counts"
test_data <- matrix(c(1489,22,793,76,521,906,13,410,42,1196),nrow=5)
rownames(test_data) <- c("EF2A","ABCD1","MEFV","BAG1","MOV10")
colnames(test_data) <- c("SampleA","SampleB")
test_data
```

Firstly, an "average" or reference sample is created that represents the counts on a typical sample in the dataset. The *geometric mean* is used rather than the *arithmetic* mean. In other words the individual counts are multiplied rather than summed and the measure should be more robust to outliers.


```{r}
psuedo_ref <- sqrt(rowProds(test_data))
psuedo_ref
```

A ratios of sample to "psuedo reference" are then calculated for each gene. We are assuming that most genes are not changing dramatically, so this ratio should be somewhere around 1.

```{r}
test_data/psuedo_ref
```

`DESeq2` defines size factors as being the *median* of these ratios for each sample (median is used so any outlier genes will not affect the normalisation). 

```{r}
norm_factors <- colMedians(test_data/psuedo_ref)
norm_factors
```

Individual samples can then normalised by dividing the count for each gene by the corresponding normalization factor.

```{r}
test_data[,1] / norm_factors[1]
```

and for the second sample...

```{r}
test_data[,2] / norm_factors[2]

```

The size factors for each sample in our dataset can be calculated using the `estimateSizeFactorsForMatrix` function.

```{r}
sf <- estimateSizeFactorsForMatrix(assay(dds))
sf
```

The estimation of these factors can also take gene-lengths into account, and this is implemented in the `estimateSizeFactors` function. Extra normalization factor data is added to the `dds` object.



```{r eval=FALSE}
dds <- estimateSizeFactors(dds)
dds
```

In preparation for differential expression DESeq2 also need a reliable estimate of the variability of each gene; which it calls *dispersion*. 

```{r eval=FALSE}
dds <- estimateDispersions(dds)
dds

```

A statistical test can then be applied. As the data are count-based and not normally-distributed a t-test would not be appropriate. Most tests are based on a *Poisson* or *negative-binomial* distribution; negative binomial in the case of `DESeq2`. Although you might not be familiar with the negative binomial, the results should be in a familiar form with fold-changes and p-values for each gene.

```{r eval=FALSE}
dds <- nbinomWaldTest(dds)
```

It may seem like there is a lot to remember, but fortunately there is one convenient function that will apply the three steps. The messages printed serve as reminders of the steps included.

```{r}
de_cellType <- DESeq(dds)
de_cellType
```



The results of the analysis are not immediately accessible, but can be obtained using the `results` function. Each row is a particular gene measured in the study (i.e. all genes in the organism being studied) and each column reports some aspect of the differential expression analysis for that gene. Note that **all genes** are reported. At this stage the gene identifiers are not very informative, something we will fix in the next section. 

```{r}
results(de_cellType)
```

# Processing the DE results using tidyverse

The output can be converted into a data frame and manipulated in the usual manner. It is recommended to use `dplyr` to manipulate the data frames with the standard set of operations detailed on the [dplyr cheatsheet](https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf)

- `select` to pick which columns to display
- `filter` to restrict the rows
- `mutate` to add new variables to the data frame
- `arrange` to order the data frame according to values of a column

However, `dpylr` does not like data frame that have rownames. We can use the `rownames_to_column` function from the `tibble` package to add an extra column that contains the Ensembl gene IDs.

The `%>%` symbol refers to the [piping operation](https://sbc.shef.ac.uk/r-online/part2.nb.html#%E2%80%9CPiping%E2%80%9D) in R, which is a way of chaining operations together. 

```{r warning=FALSE, message=FALSE}
library(dplyr)
library(tibble)
results_cellType <- results(de_cellType) %>%
  as.data.frame %>% 
  rownames_to_column("ENSEMBL")  

results_cellType
```

We can sort the rows by adjusted p-value and then print the first 10 rows.

```{r}
arrange(results_cellType, padj) %>%  
  head(n=10)
```

Or we can sort the rows and then write the resulting data frame to a file.

```{r}
arrange(results_cellType, padj) %>%
  write.csv("basal_vs_luminal_DESeq_all.csv")
```

Filtering to the differentially-expressed genes can be achieved using the `filter` function from `dplyr`.

```{r}
arrange(results_cellType, padj) %>%
  filter(padj < 0.05) %>% 
  write.csv("basal_vs_luminal_DESeq_DE.csv")
```

It is also a good idea to save the results object itself so we can re-use later.

```{r}
saveRDS(results_cellType, file="Robjects/results_cellType.rds")

```


### Exercise

<div class="exercise">
- Re-run the analysis to find differentially-expressed genes between the developmental stages *virgin* and *lactation*
- Write a csv file that contains results for the genes that have a p-value less than 0.05 and a log2 fold change more than 1, or less than -1.
- HINT: So that we don't overwrite our results so far, it may be convenient to create a new `DESeqDataSet` object for the new differential expression analysis. Check the `colData` to see which analyses can be made

```{r eval=FALSE}
dds_status <- dds
colData(dds)
design(dds_status) <- ~......
```

</div>




# Changing the direction of the contrast

In this initial analysis `DESeq2` has automatically decided which member of our sample groups to use as our baseline (`basal` in this case) so that the log2 fold changes are reported with a positive value meaning higher expression in `luminal`. If we want to change this behaviour we can change the `contrast` argument in the `results` function


```{r eval=FALSE}
## This should give the same as the table above
results(de_cellType, contrast=c("CellType","luminal","basal"))
## Changing the direction of the contrast
results(de_cellType, contrast=c("CellType","basal","luminal"))

```

If we change to performing differential expression analysis on the `Status` variable then there are various contrasts that can be made; `pregnant` vs `lactation`, `lactation` vs `virgin` etc. When the `results` function is run, the table that is displayed is for the contrast `virgin vs lactate`. The `resultsNames` function can tell us which other contrasts we can access.

For a factor with more than 2 groups, the `contrast` argument can be used to change the output. Lets say we have a factor with groups `A`, `B` and `C`:-

```{r eval=FALSE}
## this code won't run. it's just an example

## print results of comparing A to B
results(de_object, contrast = c("Group", "A","B"))

## print results of comparing A to C
results(de_object, contrast = c("Group", "A","C"))

## print results of comparing B to C
results(de_object, contrast = c("Group", "B","C"))


```


# Intersecting gene lists

A venn diagram is a common way of visualising the overlap between two genelists. We need to create a data frame where each column indicates whether each gene is differentially expressed in a particular contrast or not. To create such columns we can do a logical test on the adjusted p-values from our results tables.

The following shows a (not particularly meaningful) example.

```{r message=FALSE}
dds_status <- dds
dds_status$Status <- as.factor(dds$Status)
design(dds_status) <- ~Status
de_status <- DESeq(dds_status)
results_status <- results(de_status) %>%
  as.data.frame %>% 
  rownames_to_column("ENSEMBL")  

venn_data <- data.frame(CellType = results_cellType$padj<0.05,
                        Status = results_status$padj < 0.05)
library(limma)
vennDiagram(venn_data)
```



### Exercise

<div class="exercise">
- Use a venn diagram to visualise the overlap in the genes found to be differentially expressed in the `pregnant vs virgin` and ` lactation vs virgin` contrasts. 
- How many genes are in common?

</div>

# Analysing a subset of the data

From our initial inspection of the data we observed a large difference between `basal` and `luminal` cell types (greater than the variability between development stages). Therefore we try and compare differences in `Status` some of the biological variation could be masked by the differences in `CellType`.

One way around this would be to create different subsets for each `CellType` and analyse separately. `DESeq2` allows the dataset to be subset uses conventions that are consistent with base R data frames.

```{r}
## create subset corresponding to just basal samples
dds_basal <- dds[,which(dds$CellType == "basal")]
dds_basal$Status <- as.factor(dds_basal$Status)
design(dds_basal) <- ~Status
## perform analysis
de_basal <- DESeq(dds_basal)
```

```{r}
## create subset corresponding to just luminal samples
dds_luminal <- dds[,which(dds$CellType == "luminal")]
dds_luminal$Status <- as.factor(dds_luminal$Status)
design(dds_luminal) <- ~Status
de_luminal <- DESeq(dds_luminal)
```



# Fitting alternative models to the data

`DESEq2` allows for more complicated models to be fit to the data. For guidance on how to fit more complicated models you can consult the [DESeq2 vignette](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html), the [limma user guide](https://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf) or the Bioconductor mailing list. 

In particular, DESeq2 allows [multi-factor models](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#multi-factor-designs) which can account for other sources of variation in the data such as batches or gender.

Lets suppose that we wanted to find genes differentially-expressed between virgin and lactating individuals, but controlling for `CellType`. The main assumption being that the effect of `Status` is the same regardless of `CellType` The design for such an analysis would be:-

```{r}
dds_mf <- dds
dds_mf$Status <- as.factor(dds_mf$Status)
design(dds_mf) <- ~CellType+Status
de_mf <- DESeq(dds_mf)
results_mf <- results(de_mf,contrast=c("Status","virgin","lactation"
                                       ))
results_mf
```


### Exporting normalized counts

The `DESeq` workflow applies *median of ratios normalization* that accounts for differences in sequencing depth between samples. The user does not usually need to run this step. However, if you want a matrix of counts for some application outside of Bioconductor the values can be extracted from the `dds` object.

```{r}
dds <- estimateSizeFactors(dds) 
countMatrix <-counts(dds, normalized=TRUE)
head(countMatrix)
write.csv(countMatrix,file="normalized_counts.csv")
```



# Exercise for next time

<div class="exercise">

- Make a venn diagram for genes with adjusted p-value less than 0.05 in the basal (`de_basal`) and luminal (`de_luminal`) subsets. How many genes overlap?
- How many differentially expressed genes are found with adjusted p-value less than 0.05 in the *multi-factor* analysis? How does this compare to the number of genes in the venn diagram? What do you think is the best way to analyse these data?
- You may have noticed that some genes in the `DESeq2` output have `NA` values for p-value. Why is this? Check the "Frequently asked questions" section of the DESeq2 manual to find out. Is there an argument in the `results` function to disable this behaviour?

</div>

# Acknowledgements

**Original Authors: Belinda Phipson, Anna Trigos, Matt Ritchie, Maria Doyle, Harriet Dashnow, Charity Law**, **Stephane Ballereau, Oscar Rueda, Ashley Sawle**
Based on the course [RNAseq analysis in R](http://combine-australia.github.io/2016-05-11-RNAseq/) delivered on May 11/12th 2016 and modified by Cancer Research Uk Cambridge Centre for the [Functional Genomics Autumn School 2017](https://bioinformatics-core-shared-training.github.io/cruk-autumn-school-2017/)

## Resources and data files

This material has been created using the following resources:  

- http://www.statsci.org/smyth/pubs/QLedgeRPreprint.pdf 
- http://monashbioinformaticsplatform.github.io/RNAseq-DE-analysis-with-R/99-RNAseq_DE_analysis_with_R.html  
- http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
- https://bioconductor.github.io/BiocWorkshops/rna-seq-data-analysis-with-deseq2.html